util/libcrypto.num

```diff
@@ -4402,20 +4402,6 @@ EVP_PKEY_set1_engine                    4347	1_1_0g	EXIST::FUNCTION:ENGINE
 DH_new_by_nid                           4348	1_1_1	EXIST::FUNCTION:DH
 DH_get_nid                              4349	1_1_1	EXIST::FUNCTION:DH
 CRYPTO_get_alloc_counts                 4350	1_1_1	EXIST::FUNCTION:CRYPTO_MDEBUG
-RAND_POOL_new                           4351	1_1_1	EXIST::FUNCTION:
-RAND_POOL_free                          4352	1_1_1	EXIST::FUNCTION:
-RAND_POOL_buffer                        4353	1_1_1	EXIST::FUNCTION:
-RAND_POOL_detach                        4354	1_1_1	EXIST::FUNCTION:
-RAND_POOL_entropy                       4355	1_1_1	EXIST::FUNCTION:
-RAND_POOL_length                        4356	1_1_1	EXIST::FUNCTION:
-RAND_POOL_entropy_available             4357	1_1_1	EXIST::FUNCTION:
-RAND_POOL_entropy_needed                4358	1_1_1	EXIST::FUNCTION:
-RAND_POOL_bytes_needed                  4359	1_1_1	EXIST::FUNCTION:
-RAND_POOL_bytes_remaining               4360	1_1_1	EXIST::FUNCTION:
-RAND_POOL_add                           4361	1_1_1	EXIST::FUNCTION:
-RAND_POOL_add_begin                     4362	1_1_1	EXIST::FUNCTION:
-RAND_POOL_add_end                       4363	1_1_1	EXIST::FUNCTION:
-RAND_POOL_acquire_entropy               4364	1_1_1	EXIST::FUNCTION:
 OPENSSL_sk_new_reserve                  4365	1_1_1	EXIST::FUNCTION:
```

Comment:
We should remember to renumber the symbols following.  To be noted, there are two symbols marked 1_1_0h, a version that hasn't been released yet, so we will need to synchronise the 1.1.0 `util/libcrypto.num` accordingly.

If noone else wants to deal with that, I can.

Reply:
With pleasure, I just didn't dare to renumber existing numbers. I can do the renumbering, if you tell me which rules I need to obey. For example:
1. Is there a rule that forbids renumbering exports which already exist in stable branches?
2. Must the number of a symbol be identical in all branches?
3. What about the symbols marked 1_1_0h? Do they have to be moved towards the beginning, in front of all symbols that are new in 1.1.1?

Reply:
You can't renumber anything in the released version, which is as good as the same as in a stable branch. The numbers should match in those branches that have the same soname, which means that with our version scheme since 1.0, the first 2 digits should match. 1.0.0, 1.0.1 and 1.0.2 all have the same soname. 1.1.0 and 1.1.1 also have the same soname, but different than the 1.0.X ones.

I guess the order of the numbers itself doesn't really matter as long as they're unique, but it might be useful have everything from 1.1.0 first, then 1.1.1. But that might not work somewhere in the future after the 1.1.1 release where we need to add a new function in 1.1.0 for security reasons.

Reply:
1. Yes, if those numbers are present in existing releases.  For the 1.1.0 branch, it means no touching those marked `1_1_0g` and older, but those marked `1_1_0h` are fine since 1.1.0h hasn't been released yet.
2. Those that co-exist both in master and in the 1.1.0 branch MUST be the same.  Since 1.1.0 broke the ABI, we also broke away all the numbers to be different from the 1.0.x series.  (*)
3. No, they only need to have the exact same number in the two branches.  Also, the last `1_1_0g` number is 4347, so you MUST NOT touch any number up to that one.

Me, I'd simply renumber 4365 and on down to become 4351 and on, and then take note of what happens with the `1_1_0h` numbers and adjust the 1.1.0 branch accordingly.

-----
(*) The reason for these numbers was Windows and is still VMS.  The way shared libraries work (and we still do that on Windows) is that each numbers works as an index to a transfer vector that points out the actual location of each named function or variable.  For ABI compatibility reasons, they must therefore be identical in all branches where we promise such compatibility.  Of course, the compatibility only works for an version upgrade, not for a version downgrade.
The next time we can throw all these numbers in the air is when we release 1.2.

Reply:
Kurt and Richard, thanks for your detailed explanations!

P.S: Yeah, I‘m familiar with module definition files, I‘m a child of both worlds, Unix/Linux and Windows (although as a programmer, I prefer the former).

Reply:
> P.S: Yeah, I‘m familiar with module definition files, I‘m a child of both worlds, Unix/Linux and Windows (although as a programmer, I prefer the former).

Cool, good to know :-)